cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -0,0 +1,1033 @@
+                         uint32_t pq_dim,
+                         uint32_t pq_bits,       // 4 <= pq_bits <= 8
+                         const uint32_t* label,  // [pq_dim, ldDataset]
+                         uint8_t* output,        // [n_rows, pq_dim]
+                         rmm::cuda_stream_view stream)
+{
+  switch (pq_bits) {
+    case 4: return ivfpq_encode_run<4>(n_rows, pq_dim, label, output, stream);
+    case 5: return ivfpq_encode_run<5>(n_rows, pq_dim, label, output, stream);
+    case 6: return ivfpq_encode_run<6>(n_rows, pq_dim, label, output, stream);
+    case 7: return ivfpq_encode_run<7>(n_rows, pq_dim, label, output, stream);
+    case 8: return ivfpq_encode_run<8>(n_rows, pq_dim, label, output, stream);
+    default: RAFT_FAIL("Invalid pq_bits (%u), the value must be within [4, 8]", pq_bits);
+  }
+}
+
+template <typename T>
+auto descending(const void* a, const void* b) -> int
+{
+  auto val_a = *reinterpret_cast<const T*>(a);
+  auto val_b = *reinterpret_cast<const T*>(b);
+  if (val_a > val_b) return -1;
+  if (val_a < val_b) return 1;
+  return 0;
+}
+
+template <typename IdxT>
+void _cuann_get_inclusiveSumSortedClusterSize(index<IdxT>& index)
+{
+  auto cluster_offsets     = index.list_offsets().data_handle();
+  auto output              = index.inclusiveSumSortedClusterSize();
+  index.numClustersSize0() = 0;
+  for (uint32_t i = 0; i < index.n_lists(); i++) {
+    output(i) = cluster_offsets[i + 1] - cluster_offsets[i];
+    if (output(i) > 0) continue;
+
+    index.numClustersSize0() += 1;
+  }
+  RAFT_LOG_DEBUG("Number of clusters of size zero: %u", index.numClustersSize0());
+  // sort
+  qsort(output.data_handle(), index.n_lists(), sizeof(IdxT), descending<IdxT>);
+  // scan
+  for (uint32_t i = 1; i < index.n_lists(); i++) {
+    output(i) += output(i - 1);
+  }
+  RAFT_EXPECTS(output(index.n_lists() - 1) == index.size(), "cluster sizes do not add up");
+}
+
+/**
+ * @brief Fill-in a random orthogonal transformation matrix.
+ *
+ * @param handle
+ * @param n_rows
+ * @param n_cols
+ * @param[out] rotation_matrix device pointer to a row-major matrix of size [n_rows, n_cols].
+ * @param rng random number generator state
+ */
+inline void make_rotation_matrix(const handle_t& handle,
+                                 uint32_t n_rows,
+                                 uint32_t n_cols,
+                                 float* rotation_matrix,
+                                 raft::random::Rng rng = raft::random::Rng(7ULL))
+{
+  common::nvtx::range<common::nvtx::domain::raft> fun_scope(
+    "ivf_pq::make_rotation_matrix(%u * %u)", n_rows, n_cols);
+  bool inplace = n_rows == n_cols;
+  auto stream  = handle.get_stream();
+  uint32_t n   = std::max(n_rows, n_cols);
+  rmm::device_uvector<float> buf(inplace ? 0 : n * n, stream);
+  float* mat = inplace ? rotation_matrix : buf.data();
+  rng.normal(mat, n * n, 0.0f, 1.0f, stream);
+  linalg::detail::qrGetQ_inplace(handle, mat, n, n, stream);
```

Comment:
Shouldn't this work with rectangular matrix as well?

Reply:
About random-rotation vs identity-transform. I assumed the option to have the latter was originally added only to avoid spending too much time on this on CPU; adding the gpu version makes the random rotation feasible for larger dim sizes. Also I'm was a little bit afraid that the identity-like transform we had before may skew the feature importance when rot_dim > dim.